models/repo.go

```diff
@@ -622,6 +622,89 @@ func (repo *Repository) GetAssignees() (_ []*User, err error) {
 	return repo.getAssignees(x)
 }
 
+func (repo *Repository) getReviewersPrivate(e Engine, doerID, posterID int64) (users []*User, err error) {
+	users = make([]*User, 0, 20)
+
+	if err = e.
+		SQL("SELECT * FROM user WHERE id in (SELECT user_id FROM access WHERE repo_id = ? AND mode >= ? AND user_id NOT IN ( ?, ?))",
+			repo.ID, AccessModeRead,
+			doerID, posterID).
+		Find(&users); err != nil {
+		return nil, err
+	}
+
+	return users, nil
+}
+
+func (repo *Repository) getReviewersPublic(e Engine, doerID, posterID int64) (users []*User, err error) {
+
+	userIDs := make([]int64, 0, 20)
+
+	if err = e.
+		SQL("SELECT user_id FROM access WHERE repo_id = ? AND mode >= ? AND user_id NOT IN ( ?, ?)",
+			repo.ID, AccessModeRead,
+			doerID, posterID).
+		Find(&userIDs); err != nil {
+		return nil, err
+	}
+
+	getNum := len(userIDs)
+
+	watcherIDs := make([]int64, 0)
+
+	if err = e.SQL("SELECT user_id FROM watch WHERE repo_id = ? AND user_id NOT IN ( ?, ?)",
+		repo.ID, doerID, posterID).Find(&watcherIDs); err != nil {
+		return nil, err
+	}
+
+	for _, watcherID := range watcherIDs {
+		isHave := false
+		for index, userID := range userIDs {
+			if index >= getNum {
+				break
+			}
+			if userID == watcherID {
+				isHave = true
+				break
+			}
+		}
+
+		if !isHave {
+			userIDs = append(userIDs, watcherID)
+		}
+	}
+
+	users = make([]*User, 0, len(userIDs))
+	if err = e.In("id", userIDs).Find(&users); err != nil {
```

Comment:
This command will likely fail if the number of users is large, as SQLite only supports 997 parameters. Why not use:

```
SELECT * from `user` WHERE id IN (
    SELECT user_id from access WHERE ...
    UNION
    SELECT user_id from watch WHERE ... )
```
...similar to what you did above?

Ref: https://www.w3schools.com/sql/sql_union.asp

Reply:
yes , I don't know it before , I will check it, Thanks.